#74: an initial implementation of the Kinesis Connector Importer (#77)
exasol/cloud-storage-etl-udfs: docs/kinesis/kinesis_connector.md

# Exasol Kinesis Connector 

Exasol Kinesis Connector provides UDF scripts which allow users import data
from a Kinesis Stream to an Exasol table.

## Prerequisites

* A running Exasol cluster with a version 6.0 or later.
* An AWS account with all necessary permissions to read from a Kinesis Stream.
* An AWS Kinesis producer which sends the data to a stream in the valid JSON
 format. 
 An example of a valid JSON string: 
 `{"sensorId": 17,"currentTemperature": 147,"status": "WARN"}`.
 An example of an invalid JSON format:
 `{"sensorId": 17,"currentTemperature": 147,"status": "WARN",}`. The Exasol
 Kinesis Connector will not parse this string correctly.

## Deployment

> Please refer to the [Deployment Guide](../deployment_guide.md) till the section

Author: fixed

`Upload the JAR file to the bucket` (inclusive) to prepare a bucket and a JAR.

## Preparing a table for data

To store the data from a stream we need a table. 
The columns in the Exasol table have to imitate the data types of the data
 stored in the stream and also be in the exact same order.
 
The table also requires two additional columns to store the Kinesis metadata:  
    `kinesis_shard_id VARCHAR(2000), 
     kinesis_sequence_number VARCHAR(2000)`
These two columns must be in the end of the table.
 
For example, we have a stream with the following data: 
 `{"sensorId": 17,"currentTemperature": 147,"status": "WARN"}`

> We create a table for this stream:

Reviewer: ```suggestion
For example, given the following stream data, `{"sensorId": 17,"currentTemperature": 147,"status": "WARN"}`, the Exasol table should be as below:
```

```sql
OPEN SCHEMA <schema_name>;

CREATE OR REPLACE TABLE <schema_name>.<table_name> 
    (sensorId DECIMAL(18,0), 
    currentTemperature DECIMAL(18,0), 
    status VARCHAR(100), 
-- Required for importing data from Kinesis
    kinesis_shard_id VARCHAR(2000), 
    kinesis_sequence_number VARCHAR(2000));

> 

Reviewer: To be consistent with the Kafka, could you please move these two extra columns as first two columns of table? It is helpfule to make them visible immediately as additional required columns.

Author: We decided not to do it

```

## Create ETL UDFs Scripts

Create the following UDF scripts. Please do not change the names of the scripts.

```sql
--/
CREATE OR REPLACE JAVA SET SCRIPT READ_KINESIS_SHARDS_METADATA (...) EMITS (kinesis_shard_id VARCHAR(2000)) AS
  %scriptclass com.exasol.cloudetl.kinesis.KinesisShardsMetadataReader;

>   %jar /buckets/bfsdefault/kinesis/cloud-storage-etl-udfs-0.6.2-SNAPSHOT.jar;

Reviewer: ```suggestion
  %jar /buckets/bfsdefault/kinesis/cloud-storage-etl-udfs-<VERSION>.jar;
```

Also change the ones below.

/
;
  
--/
CREATE OR REPLACE JAVA SET SCRIPT IMPORT_KINESIS_SINGLE_SHARD_DATA (...) EMITS (...) AS
  %scriptclass com.exasol.cloudetl.kinesis.KinesisShardDataImporter;
  %jar /buckets/bfsdefault/kinesis/cloud-storage-etl-udfs-0.6.2-SNAPSHOT.jar;
/
;
    
--/
CREATE OR REPLACE JAVA SET SCRIPT GENERATE_KINESIS_IMPORT_QUERY (...) EMITS (...) AS
  %scriptclass com.exasol.cloudetl.kinesis.KinesisImportQueryGenerator;
  %jar /buckets/bfsdefault/kinesis/cloud-storage-etl-udfs-0.6.2-SNAPSHOT.jar;
/

> ; 

Reviewer: Rename the UDF name into: `KINESIS_METADATA`, `KINESIS_IMPORT` and `KINESIS_PATH`.

Author: It's very misleading in my eyes that I call `KINESIS_PATH` script to run an import command. I used these names first, but they bothered me so much that I renamed them to make it easier to understand what each of them does

# Exasol Kinesis Connector 

Exasol Kinesis Connector provides UDF scripts that allow users to import data
from [Kinesis Streams][kinesis-streams] to an Exasol table.

## Prerequisites

* A running Exasol cluster with a version 6.0 or later.
* An AWS account with all necessary permissions to read from a Kinesis Stream.
* An AWS Kinesis producer which sends the data to a stream in the valid JSON
 format. 
 An example of a valid JSON string: 
 `{"sensorId": 17,"currentTemperature": 147,"status": "WARN"}`.
 An example of an invalid JSON format:
 `{"sensorId": 17,"currentTemperature": 147,"status": "WARN",}`. The Exasol

>  Kinesis Connector will not parse this string correctly.

Reviewer 2: ```suggestion
 An example of a valid JSON string: 
 
    {"sensorId": 17,"currentTemperature": 147,"status": "WARN"}`
  
  
See an example of an invalid JSON format. Note the trailing comma.
 
     {"sensorId": 17,"currentTemperature": 147,"status": "WARN",}`
     
 The Exasol Kinesis Connector will not parse this string correctly.
```

## Deployment

> Please refer to the [deployment guide](../deployment_guide.md) till the section

Reviewer 2: ```suggestion
Please refer to the [deployment guide](../deployment_guide.md) up to untill the section
```

`Upload the JAR file to the bucket` (inclusive) to prepare a bucket and a JAR.

## Preparing a Table for Data

To store the data from a stream, we need a table. 
The columns in the Exasol table have to imitate the data types of the data
 stored in the stream and also be in the exact order.
 
The table also requires two additional columns to store the Kinesis metadata:  
    `KINESIS_SHARD_ID VARCHAR(2000), 
     SHARD_SEQUENCE_NUMBER VARCHAR(2000)`

> These two columns must be at the end of the table.

Reviewer 2: ```suggestion
The table also requires two additional columns to store the Kinesis metadata: 

    KINESIS_SHARD_ID VARCHAR(2000), 
    SHARD_SEQUENCE_NUMBER VARCHAR(2000)

These two columns must be at the end of the table.
```

 
For example, given the following stream data, 
`{"sensorId": 17,"currentTemperature": 147,"status": "WARN"}`, 

> the Exasol table should be as below:

Reviewer 2: ```suggestion
As an example, take the following stream of data:

    {"sensorId": 17,"currentTemperature": 147,"status": "WARN"}
     
The  corresponding Exasol table should look like:
```

```sql
CREATE SCHEMA <schema_name>;

CREATE OR REPLACE TABLE <schema_name>.<table_name> 
    (SENSORID DECIMAL(18,0), 
    CURRENTTEMPERATURE DECIMAL(18,0), 
    STATUS VARCHAR(100), 
-- Required for importing data from Kinesis
    KINESIS_SHARD_ID VARCHAR(130), 
    SHARD_SEQUENCE_NUMBER VARCHAR(2000));

```

## Create ETL UDFs Scripts

Create the following UDF scripts. Please do not change the names of the scripts.

```sql
--/
CREATE OR REPLACE JAVA SET SCRIPT KINESIS_METADATA (...) 
EMITS (KINESIS_SHARD_ID VARCHAR(130), SHARD_SEQUENCE_NUMBER VARCHAR(2000)) AS
  %scriptclass com.exasol.cloudetl.kinesis.KinesisShardsMetadataReader;

>   %jar /buckets/bfsdefault/kinesis/cloud-storage-etl-udfs-<VERSION>.jar;

Reviewer 2: You could have our `version` script inject the latest versions here automatically like in the Virtual Schemas. This would be more convenient for the end users.

Reviewer: I think this will not work. The version is obtained using `git describe`. So every commit has a version, e.,g `tag-commitNumberSince-sha`. 

Also, I think it will be too much. User can obtain the latest released jar from github releases.

/
;
  
--/
CREATE OR REPLACE JAVA SET SCRIPT KINESIS_IMPORT (...) EMITS (...) AS
  %scriptclass com.exasol.cloudetl.kinesis.KinesisShardDataImporter;
  %jar /buckets/bfsdefault/kinesis/cloud-storage-etl-udfs-<VERSION>.jar;
/
;
    
--/
CREATE OR REPLACE JAVA SET SCRIPT KINESIS_PATH (...) EMITS (...) AS
  %scriptclass com.exasol.cloudetl.kinesis.KinesisImportQueryGenerator;
  %jar /buckets/bfsdefault/kinesis/cloud-storage-etl-udfs-<VERSION>.jar;
/
; 
```

## Import Data

You can provide AWS credentials using:
1. UDF parameters 

> 1. [Exasol named connection object][exa-connection]. 

Reviewer 2: ```suggestion
You can provide AWS credentials using:

1. UDF parameters 
1. [Exasol named connection object][exa-connection]. 
```

We highly **recommend** using connection objects to provide credentials 

> to UDFs so that credentials do not display in audit logs.

Reviewer 2: ```suggestion
We highly recommend you **use connection objects** to provide credentials 
to UDFs so that credentials do not display in audit logs.
```

This way the emphasis is on the important piece.

> ### Using Connection Object

Reviewer 2: ```suggestion
### Using the Connection Object
```

First, create a named connection object and encode credentials as key-value
pairs separated by semicolon (`;`).

`AWS_SESSION_TOKEN` is an optional property. You can remove it from the string
if you don't need it. 

```sql
CREATE OR REPLACE CONNECTION KINESIS_CONNECTION
TO ''
USER ''

> IDENTIFIED BY 'AWS_ACCESS_KEY=<AWS_ACCESS_KEY>;AWS_SECRET_KEY=<AWS_SECRET_KEY>;AWS_SESSION_TOKEN=<AWS_SESSION_TOKEN>';

Reviewer 2: Please mention where to get those credentials briefly. E.g. by pointing to the corresponding AWS documentation.

Below is a table with all properties. Maybe just add links there.

Author: Added access keys as a requirement and also inserted a link

```

Run an import query.
 
```sql
IMPORT INTO <table_name>
FROM SCRIPT KINESIS_PATH WITH

>   TABLE_NAME     = '<table_name>'

Reviewer 2: Please either line up the equal signs and values or use one space to each side of the assignment. This makes me nervous. :-)

Author: done

  CONNECTION_NAME  = 'KINESIS_CONNECTION'
  STREAM_NAME    = '<stream_name>'
  REGION    = '<region>'
;
``` 

### Without Connection Object

Provide credentials as properties and run an import query.

**Attention! Providing credentials via propeties is 

> deprecated and will be removed in future releases.**

Reviewer 2: Add a deprecation warning to the central README too. Prominently, so that users have no excuse to not read it.

Author: done

Author: We will do it later
